### PR TITLE
Update status handling

### LINKED ISSUE #23: Move the TF StateAnnoation to a Status field on CRD (under top level tfbridgestate object)

### PR DESCRIPTION
**Pull request by stuartleeks**
* Add `make build` command
* Save state via `Patch` rather than `Update` to make resilient to changes in the resource between getting and saving (can be quite a long period due to resource creation time)
* Make `status` a subresource on CRDs (fixes #23)
* Add `_tfoperator` property on `status` to store `tfState` and `lastAppliedGeneration` instead of using annotations
* Add `kind` and `gen` values to controller logger
* Pass context and logger into `Reconcile` calls and use for appropriate operations

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,3 +1,5 @@
+build: 
+	go build .
```
Comment (lawrencegripper) [inline Makefile:2]: Kinda funny we didn't have this!

File: controller.go [hunk 2]
```diff
@@ -49,24 +49,20 @@ func (r *controller) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 		if errors.IsNotFound(err) {
 			return ctrl.Result{}, nil
 		}
-		log.Error(err, "Failed getting resource", "req", req.NamespacedName)
+		log.Error(err, "Failed getting resource")
 		return ctrl.Result{}, fmt.Errorf("Failed getting resource %q %w", req.NamespacedName.String(), err)
 	}
-	gen := resource.GetGeneration()
-	log.Info("*** Reconciler called", "namespace", resource.GetNamespace(), "kind", resource.GetKind(), "name", resource.GetName(), "generation", gen)
+	log = log.WithValues("kind", resource.GetKind(), "gen", resource.GetGeneration())
```
Comment (lawrencegripper) [inline controller.go:55]: Nice

File: create.go [hunk 3]
```diff
@@ -190,6 +201,9 @@ func installCRDs(resources []spec.Schema, providerName, providerVersion string)
 				Validation: &apiextensionsv1beta1.CustomResourceValidation{
 					OpenAPIV3Schema: &jsonSchemaProps,
 				},
+				Subresources: &apiextensionsv1beta1.CustomResourceSubresources{
+					Status: &apiextensionsv1beta1.CustomResourceSubresourceStatus{},
+				},
 			},
```
Comment (lawrencegripper) [inline create.go:207]: Ooooh shiny

File: main.go [hunk 4]
```diff
@@ -11,18 +11,18 @@ import (
...
 )
 
 func main() {
 	provider := getInstanceOfAzureRMProvider()
 
 	// Example talking to the Azure resources using the provider
-	configureProvider(provider)
-	// exampleChangesToResourceGroup(provider)
+	log := ctrl.Log.WithName("main")
+	configureProvider(log, provider)
```
Comment (lawrencegripper) [inline main.go:22]: Nice

File: reconcile.go [hunk 5]
```diff
@@ -1,256 +1,216 @@
...
 		configValue = &v
 	} else {
-		ensureFinalizer(crd)
+		if err := r.ensureFinalizer(ctx, log, crd); err != nil {
+			return reconcileLogError(log, fmt.Errorf("Error adding finalizer: %s", err))
+		}
 
 		// Get the spec from the CRD
-		spec := crd.Object["spec"].(map[string]interface{})
+		spec, gotSpec, err := unstructured.NestedMap(crd.Object, "spec")
```
Comment (lawrencegripper) [inline reconcile.go:54]: oh yeah this is nicer with the `NestedMap` helpers

File: reconcile.go [hunk 6]
```diff
@@ -1,256 +1,216 @@
...
 		}
 	}
-	metadata["finalizers"] = updatedFinalizers
-	resource.Object["metadata"] = metadata
+	if !foundFinalizer {
+		log.Info("Finalizer - not found on remove")
+		return fmt.Errorf("Finalizer not found on remove")
+	}
+	resource.SetFinalizers(updatedFinalizers)
+	if err := r.client.Patch(ctx, resource, client.MergeFrom(copyResource)); err != nil {
```
Comment (lawrencegripper) [inline reconcile.go:161]: I really like this patch approach using the deepcopy

File: reconcile.go (hunk 5 above)
Comment (stuartleeks) [reply reconcile.go:54]: Yeah - they were a good call. Thanks for pointing them out!

File: reconcile.go (hunk 6 above)
Comment (stuartleeks) [reply reconcile.go:161]: Yeah, it feels more targetted. Apart from `finalizers`, I think all the updates are on `status` now which should reduce the chance of conflicts now that it's a subresource.

**Pull request merged**